posts/ebpf-and-kernel-modules-whats-the-difference.md

```diff
@@ -0,0 +1,115 @@
+
+- eBPF
+	- eBPF programs that are compiled with an eBPF developer library that support [Compile-Once Run-Everywhere (CO-RE)](https://facebookmicrosites.github.io/bpf/blog/2020/02/19/bpf-portability-and-co-re.html), will be able to continue to work on kernel updates even if kernel data structures or other aspects of the kernel change. This is because of abstraction layers and the design of eBPF libraries take when building the eBPF program as well as abstraction layers present in the run-time.
+	- One thing to note is that there are eBPF development libraries that do not take advantage of CO-RE. One such library is BCC, which is a library that many eBPF programmers start in because of the greater simplicity.
+	- Production projects will be more incentivized to write eBPF programs using the more complex, but more portable CO-RE libraries. One CO-RE library is [libbpf](https://github.com/libbpf/libbpf) for C.
+
+A takeaway from this is that development of eBPF programs and the linux kernel can be developed independently of each other, whereas a kernel module is very dependent on linux kernel development updates.
+
+
+# Learning More
+### eBPF
+To learn about eBPF development, I would recommend starting with the BCC Python library. This is a user-space library that compiles and runs eBPF C programs. It can be installed by following [these official instructions](https://github.com/iovisor/bcc/blob/master/INSTALL.md).
+
+To write a simple hello world program, [this article](https://rdelfin.com/blog/post_ebpf_002) should get you started. You can ignore the installation instructions as following the official BCC install instructions mentioned above will suffice. If you run into macro redefinition errors while going through this article, you can reference [this github issue](https://github.com/iovisor/bcc/issues/3366#issuecomment-1258054405).
+
+If you want to go deeper into advanced topics, I recommend the book [Learning eBPF by Liz Rice](https://isovalent.com/learning-ebpf/)
+
+### Linux Kernel Modules
+To learn about kernel module development, [this guide](https://sysprog21.github.io/lkmpg/) is very useful. It provides a simple hello world program at the beginning and then delves into more advanced topics.
+
+
+# Conclusion
+Although eBPF is more rigid and provides fewer use-cases than kernel modules, it is generally easier to work with as there are higher-level libraries for interacting with kernel functionality than kernel modules do. It may be a good option to go with eBPF if the kernel you're working with supports the needed eBPF features (eBPF features are constantly being updated and may not be supported by older kernel versions) and if your use-case involves Networking, Security, or Observability. If your use-case is outside of that, you will probably have to stick with kernel modules for now. Time will tell if eBPF starts including more features to overtake use-cases that were traditionally reserved for kernel modules.
```

Comment:
add a call to action like - email sales if you need help implementing an eBPF module in your organization!

Reply:
added